### PR TITLE
New request fields must be defaulted or optional. WA-98 QA-1031

### PR DESCRIPTION
**Pull request by kshakir**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by kshakir** [comment]
Note: No linked PR needed for firecloud-orchestration. It turns out instead of using booleans with defaults, [orch was already using tristates and avoided the issue](https://github.com/broadinstitute/firecloud-orchestration/blob/a1b92aeec4b6f83135fc807d3a9fbb39e68900ae/src/main/scala/org/broadinstitute/dsde/firecloud/model/Workspace.scala#L82).

**Review by salonishah11 — APPROVED** [review]

**Comment by salonishah11** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala [hunk 1]
```diff
@@ -333,7 +333,57 @@ class ExecutionJsonSupport extends JsonSupport {
...
 
-  implicit val SubmissionRequestFormat = jsonFormat8(SubmissionRequest)
+  /*
+  Copied from:
+  - https://www.codersbistro.com/blog/default-values-with-spray-json-marshalling/
+  - https://gist.github.com/agemooij/f2e4075a193d166355f8
+  If you find a better way of implementing this (*cough* circe?) please add a link
+
+  NOTE:
+  - Until REST API endpoint versioning across the many repositories / consumers is figured out, DO NOT DELETE ANY FIELDS
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:343]: I... don't think this is true. If I delete a field from the request and a client continues sending it, that will be fine. In general superfluous fields are non-problematic.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala [hunk 2]
```diff
@@ -333,7 +333,57 @@ class ExecutionJsonSupport extends JsonSupport {
...
+      IN THE REQUEST, only deprecate and/or default!
+  - We are operating under an assumption for clients: even if the API version doesn't change, clients will hopefully
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:345]: This is more than a hope, it is the general contract under which we inter-operate microservices

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala [hunk 3]
```diff
@@ -333,7 +333,57 @@ class ExecutionJsonSupport extends JsonSupport {
...
+        entityType = fields.get("entityType").map(_.convertTo[String]),
+        entityName = fields.get("entityName").map(_.convertTo[String]),
+        expression = fields.get("expression").map(_.convertTo[String]),
+        useCallCache = fields("useCallCache").convertTo[Boolean],
+        // All new fields below this line MUST have defaults or be wrapped in Option[]!
+        // https://broadworkbench.atlassian.net/browse/QA-1031
+        // NOTE: All fields are optional in firecloud-orchestration's copy of this class,
+        // so in that project one can just adjust the function call to jsonFormat*N*.
+        deleteIntermediateOutputFiles = fields.get("deleteIntermediateOutputFiles").fold(false)(_.convertTo[Boolean]),
+        // All new fields above this line MUST have defaults or be wrapped in Option[]!
```
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala]: This is kind of weird... what if I add a new optional case class member after `workflowFailureMode`?

**Review by aednichols — APPROVED** [review]
Pre-merge criterion:
- Make sure that FiaB started with the correct version of Rawls and that it successfully ran its test workflow in test `to launch workflows if they have can-compute permission`
- This is because there is recent concern around FiaBs launching with the right Docker image versions

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala (hunk 1 above)
Comment (kshakir) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:343]: After today's chaos with curl failing, I no longer believe in "in general". I have ZERO confidence that there does not exist a test repo floating out there with a:

```console
$ curl http://rawls/ | jq .someFieldThatYouWouldNotExpect
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala (hunk 2 above)
Comment (kshakir) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:345]: For now. We keep changing things, and there is no actual SemVer for our webservices. I'm leaving this a warning after getting burned.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala (hunk 3 above)
Comment (kshakir) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala]: Will surround the entire block with the comment, but due laser focus I've run into PRs that don't look at lines of code more than +/-3 lines away. 🤞 someone actually reads the docs and doesn't trip over this again.

Speaking of which, there will be an incoming ticket to add doc pointers over to a to-be-created confluence doc.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala (hunk 1 above)
Comment (kshakir) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:343]: Edit: Whoops, sorry, replied too fast and thought you were commenting on the response.

As for the request changes, this is more a warning for editing-in-the-multi-repo. To me this system is very fragile and undocumented at the moment, and we have not identified all of the ripples in the system when deleting input fields.

Yes, _this_ heart-of-rawls code might work, but you may end up breaking something else, again like a bash-function-to-a-curl call in another repo. (PR still incoming to fix one of those projects...)

**Comment by kshakir** [comment]
🤔 

> Make sure that FiaB started with the correct version of Rawls

In [this log](https://fc-jenkins.dsp-techops.broadinstitute.org/job/fiab-start/37208/consoleFull) message appears to point to the correct commit at least:

```
14:22:56 Pulling rawls-app (broadinstitute/rawls:ks_delete_intermediate_output_files_default)...
14:22:57 ks_delete_intermediate_output_files_default: Pulling from broadinstitute/rawls
14:23:01 Digest: sha256:aef33c1f6a25220d5d1c2fcf3db7ec35ca190593e40cc473de05c3d659fefd95
```

I'll add a ticket to ping `/version` in swatomation though... as I don't see `gitHash` output in the current logs during startup, though we do output the responses of `/status`.

> and that it successfully ran its test workflow in test to launch workflows if they have can-compute permission

In [this log](https://fc-jenkins.dsp-techops.broadinstitute.org/job/rawls-fiab-test-runner/19818/consoleFull) there is:

```
[info]   should allow writers
[info]   - to launch workflows if they have can-compute permission (16 seconds, 897 milliseconds)
```

🤞

**Comment by aednichols** [comment]
Thank you for jumping on this @kshakir !

File: core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala (hunk 1 above)
Comment (breilly2) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala:343]: I think ignoring a field that previously had some effect, thereby providing different behavior than the caller expected, would be unfriendly. It's worth always thinking through the future consequences of any public API change.

**Comment by breilly2** [comment]
➕ Thank you @kshakir !

**Pull request merged**